llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp

```diff
@@ -1649,12 +1649,13 @@ void DAGTypeLegalizer::ExpandFloatRes_FCEIL(SDNode *N,
 
 void DAGTypeLegalizer::ExpandFloatRes_FCOPYSIGN(SDNode *N,
                                                 SDValue &Lo, SDValue &Hi) {
-  ExpandFloatRes_Binary(N, GetFPLibCall(N->getValueType(0),
-                                        RTLIB::COPYSIGN_F32,
-                                        RTLIB::COPYSIGN_F64,
-                                        RTLIB::COPYSIGN_F80,
-                                        RTLIB::COPYSIGN_F128,
-                                        RTLIB::COPYSIGN_PPCF128), Lo, Hi);
+
```

Comment:
Did this never handle the case where the 2 operands have different types?

Reply:
That does in fact seem to be the case, yes.

Reply:
I'm not sure why it would handle that case, though? We're in LLVM IR. The floats should be of the same type. Instructions like `fadd` and such demand it. The language reference always uses identical types for this intrinsic, as presented here: https://llvm.org/docs/LangRef.html#llvm-copysign-intrinsic

That is presumably for a reason? If the frontend wants to do type promotion, that is a C-like language semantic that clang can apply. The response that LLVM should have to these being different types is `report_fatal_error`, isn't it?

Reply:
And if it is indeed erroneous, it should have been handled already, I would think.

Reply:
This is not in LLVM IR. The IR intrinsic requires both types to match (and I consider this a flaw in the intrinsic). ISD::FCOPYSIGN does not, and the [combiner forms them](https://github.com/llvm/llvm-project/blob/f0ed31ce4b63a5530fd1de875c0d1467d4d2c6ea/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L17871)

So it's not erroneous. The type of the second operand does not matter

Reply:
I guess we arrive back at "this code doesn't have to worry about handling the operands being of different sizes" but the other way around, then.